apps/src/lab2/views/ExtraLinksModal.tsx

```diff
@@ -83,6 +102,144 @@ const ExtraLinksModal: React.FunctionComponent<ExtraLinksModalProps> = ({
+        true,
+        {contentType: 'application/json;charset=UTF-8'}
+      );
+      setFeaturedProjectStatus(FeaturedProjectStatus.bookmarked);
+    } catch (e) {
+      console.log('Error bookmarking project', e);
+    }
+  };
+
+  const renderCloneLevel = (
+    <div>
+      <Button
+        size={Button.ButtonSize.small}
+        color={Button.ButtonColor.purple}
+        onClick={() => setShowCloneField(!showCloneField)}
+        text={showCloneField ? 'Cancel Clone' : 'Clone'}
+      />
+      {showCloneField && (
+        <div>
+          {'New level name: '}
+          <input
+            type="text"
+            value={clonedLevelName}
+            onChange={event => setClonedLevelName(event.target.value)}
+          />
+          <Button
+            onClick={handleClone}
+            text={'Clone'}
+            size={Button.ButtonSize.small}
+          />
+          {cloneError && (
+            <p className={moduleStyles.errorMessage}>{cloneError}</p>
+          )}
+        </div>
+      )}
+    </div>
+  );
+
+  const renderDeleteLevel = (
+    <div>
+      <Button
+        size={Button.ButtonSize.small}
+        text={showDeleteConfirm ? 'Cancel Delete' : 'Delete'}
+        color={Button.ButtonColor.red}
+        onClick={() => setShowDeleteConfirm(!showDeleteConfirm)}
+      />
+      {showDeleteConfirm && (
+        <div>
+          {'Are you sure you want to delete this level? '}
+          <Button
+            onClick={handleDelete}
+            text={'Confirm Delete'}
+            size={Button.ButtonSize.small}
+          />
+          {deleteError && (
+            <p className={moduleStyles.errorMessage}>{deleteError}</p>
+          )}
+        </div>
+      )}
+    </div>
+  );
+
+  const renderScriptLevelPathLinks = (
+    <>
+      <StrongText>
+        This level is in{' '}
+        {Object.entries(levelLinkData.script_level_path_links).length} scripts:
+      </StrongText>
+      <ul>
+        {levelLinkData.script_level_path_links.map((link, index) => (
+          <li key={index}>
+            <a href={'/s/' + link.script}>{link.script}</a> as{' '}
+            <a href={link.path}>{link.path}</a>
+          </li>
+        ))}
+      </ul>
+    </>
+  );
+
+  const renderRemixAncestry = (remixList: string[]) => {
+    if (remixList.length === 0) {
+      return <li>Not a remix.</li>;
+    }
+    return remixList.map((link: string, index: number) => (
+      <li key={index}>
+        <a href={link}>{link}</a>
+      </li>
+    ));
+  };
+
+  const renderFeaturedProjectInfo = () => {
+    if (featuredProjectStatus === 'n/a') {
+      return (
+        <>
+          <div>Not a featured project</div>
+          <Button
+            size={Button.ButtonSize.small}
+            color={Button.ButtonColor.purple}
+            onClick={onBookmark}
+            text={'Bookmark as featured'}
+          />
+        </>
+      );
+    }
+    return <div>Featured project status: {featuredProjectStatus}</div>;
+  };
```

Comment:
Totally nonblocking, just my preferred syntactic sugar...

For all of these `render*` functions, they're fine as is, but stylistically I kinda prefer making them actual components and referencing them that way. The change is simple:

```
renderFeaturedProjectInfo -> FeaturedProjectInfo

and then
 { renderFeaturedProjectInfo()} => <FeaturedProjectInfo />
```

(or whatever component name you want.

I'd consider the function calls a slightly older style, and it feels more react-ish as components.

Then, just other notes for the other components -
- The ones that return static markup can also be wrapped into components in this manner - just have them be a function that returns something (probably preferably a memoized function to ensure no re-renders)
- The ones taking args should be tweaked so the args come in as props
- I see that `renderProjectValidatorData` is wrapped up in a conditional where it checks `projectLinkData` and `isStandaloneProject`. I might add those in as additional props and have the component return `null` if either of them are false before falling into the rendering.

But, again, totally just personal preference so only bother to update if it looks better to you too. 😄

Reply:
Thanks Jim - this makes sense and very appreciated! Follow-up work is to add Report abuse info/links as well as blockly-related buttons so I will log this suggested refactor as part of the follow-up.

Reply:
Hey Jim - doing some refactoring based on other comments so will also incorporate your feedback as well. Thanks!